nlohmann/json
# Add CI build testing JSON as a bazel dependency.

In order to make the Bazel build actually useful, it should build the unit tests (future PR) or test that the library can be used as a dependency in a Bazel project (this PR).

To test it locally:

```shell
mkdir /tmp/build_output
docker run -e USER="1000" -u="1000" -v $PWD:$PWD -v /tmp/build_output:/tmp/build_output -w $PWD l.gcr.io/google/bazel:0.28.1 --output_user_root=/tmp/build_output build --test_output=errors --verbose_failures=true --keep_going -- //...:all
```

Repeating my previous comments:

Note that the WORKSPACE file is required, otherwise Bazel will complain.

My previous approach was to list the include files in CMakeLists.txt then export it to Bazel. However, since we never use those include files anywhere else, I cut out the middleman. ~Developers will still have to remember to update nlohmann_json.bzl if a new header file is added.~ I was wrong about not using recursive glob. Re-reading the docs, it appears that it is talking about the pattern of combining recursive globbing with exclude, not recursive globbing in general.

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for these kind of bugs). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](http://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.
